File: examples/create_pipe_example.py
"""
Virtual Audio Driver - Python Named Pipe Example

This example demonstrates how to create a named pipe for the Virtual Audio Driver
to connect to. This shows the proper way to create and manage the named pipe
that the kernel driver will open.

Requirements:
    pip install pywin32 numpy

Usage:
    python create_pipe_example.py [pin_number]
"""

import sys
import time
import math
import struct
from typing import Optional

try:
    import win32pipe
    import win32file
    import numpy as np
except ImportError:
    print("Required packages not installed. Please run:")
    print("pip install pywin32 numpy")
    sys.exit(1)

# Audio constants
SAMPLE_RATE = 44100
CHANNELS = 2
BITS_PER_SAMPLE = 16
BYTES_PER_SAMPLE = BITS_PER_SAMPLE // 8

class VirtualMicPipeServer:
    def __init__(self, pin_number: int = 0):
        self.pipe_name = f"\\\\.\\pipe\\VirtualMicInput_{pin_number}"
        self.pipe_handle: Optional[int] = None
        self.is_running = False
        
    def start(self) -> bool:
        """Create and start the named pipe server"""
        try:
            # Create the named pipe
            self.pipe_handle = win32pipe.CreateNamedPipe(
                self.pipe_name,
                win32pipe.PIPE_ACCESS_OUTBOUND,
                win32pipe.PIPE_TYPE_BYTE | win32pipe.PIPE_READMODE_BYTE | win32pipe.PIPE_WAIT,
                1,      # MaxInstances
                8192,   # OutBufferSize
                8192,   # InBufferSize
                0,      # DefaultTimeOut
                None    # Security attributes
            )
            
            print(f"Named pipe created successfully: {self.pipe_name}")
            print("Virtual Audio Driver can now connect to this pipe.")
            self.is_running = True
            return True
            
        except Exception as e:
            print(f"Failed to create named pipe: {e}")
            return False
    
    def stop(self):
        """Stop the pipe server and cleanup"""
        self.is_running = False
        if self.pipe_handle:
            win32file.CloseHandle(self.pipe_handle)
            self.pipe_handle = None
    
    def send_sine_wave(self, frequency: float, duration: float) -> bool:
        """Send a sine wave of specified frequency and duration"""
        if not self.pipe_handle or not self.is_running:
            return False
        
        try:
            # Generate samples
            total_samples = int(SAMPLE_RATE * duration)
            t = np.linspace(0, duration, total_samples, dtype=np.float32)
            
            # Generate sine wave
            wave = np.sin(2 * np.pi * frequency * t) * 16000  # Amplitude of 16000
            wave = wave.astype(np.int16)
            
            # Convert to stereo
            stereo_wave = np.zeros((total_samples * CHANNELS,), dtype=np.int16)
            stereo_wave[0::2] = wave  # Left channel
            stereo_wave[1::2] = wave  # Right channel
            
            # Convert to bytes
            audio_data = stereo_wave.tobytes()
            
            # Send to pipe
            win32file.WriteFile(self.pipe_handle, audio_data)
            
            print(f"Sent {len(audio_data)} bytes ({duration:.1f}s of {frequency:.0f}Hz sine wave)")
            return True
            
        except Exception as e:
            print(f"Failed to send sine wave: {e}")
            return False
    
    def send_silence(self, duration: float) -> bool:
        """Send silence for specified duration"""
        if not self.pipe_handle or not self.is_running:
            return False
        
        try:
            # Generate silence (zeros)
            total_samples = int(SAMPLE_RATE * duration * CHANNELS)
            silence = np.zeros(total_samples, dtype=np.int16)
            audio_data = silence.tobytes()
            
            # Send to pipe
            win32file.WriteFile(self.pipe_handle, audio_data)
            
            print(f"Sent {duration:.1f}s of silence ({len(audio_data)} bytes)")
            return True
            
        except Exception as e:
            print(f"Failed to send silence: {e}")
            return False
    
    def send_white_noise(self, duration: float, amplitude: float = 0.1) -> bool:
        """Send white noise for specified duration"""
        if not self.pipe_handle or not self.is_running:
            return False
        
        try:
            # Generate white noise
            total_samples = int(SAMPLE_RATE * duration)
            noise = np.random.normal(0, amplitude, total_samples) * 16000
            noise = np.clip(noise, -32767, 32767).astype(np.int16)
            
            # Convert to stereo
            stereo_noise = np.zeros((total_samples * CHANNELS,), dtype=np.int16)
            stereo_noise[0::2] = noise  # Left channel
            stereo_noise[1::2] = noise  # Right channel
            
            # Convert to bytes
            audio_data = stereo_noise.tobytes()
            
            # Send to pipe
            win32file.WriteFile(self.pipe_handle, audio_data)
            
            print(f"Sent {duration:.1f}s of white noise ({len(audio_data)} bytes)")
            return True
            
        except Exception as e:
            print(f"Failed to send white noise: {e}")
            return False

def print_usage():
    print("Usage: python create_pipe_example.py [pin_number]")
    print("  pin_number: Virtual microphone pin number (default: 0)")
    print("")
    print("This program creates a named pipe and sends test audio data.")
    print("Make sure the Virtual Audio Driver is installed and running.")

def main():
    pin_number = 0
    
    # Parse command line arguments
    if len(sys.argv) > 1:
        if sys.argv[1] in ["/?" , "-h", "--help"]:
            print_usage()
            return 0
        try:
            pin_number = int(sys.argv[1])
            if pin_number < 0:
                print("Invalid pin number. Using default: 0")
                pin_number = 0
        except ValueError:
            print("Invalid pin number. Using default: 0")
            pin_number = 0
    
    print("Virtual Audio Driver - Python Named Pipe Example")
    print("=" * 50)
    print(f"Using microphone pin: {pin_number}")
    print()
    
    pipe_server = VirtualMicPipeServer(pin_number)
    
    # Start the pipe server
    if not pipe_server.start():
        print("Failed to start pipe server")
        return 1
    
    print("Pipe server started. Sending test audio...")
    print("Press Ctrl+C to stop.")
    print()
    
    try:
        cycle = 0
        while True:
            cycle += 1
            print(f"\n--- Cycle {cycle} ---")
            
            # Send 1 second of 440Hz (A4)
            pipe_server.send_sine_wave(440.0, 1.0)
            time.sleep(0.1)
            
            # Send 1 second of 880Hz (A5)
            pipe_server.send_sine_wave(880.0, 1.0)
            time.sleep(0.1)
            
            # Send some white noise
            pipe_server.send_white_noise(0.5, 0.1)
            time.sleep(0.1)
            
            # Send 1 second of silence
            pipe_server.send_silence(1.0)
            time.sleep(0.1)
            
            # Send a sweep from 200Hz to 2000Hz
            for freq in [200, 400, 800, 1200, 1600, 2000]:
                pipe_server.send_sine_wave(freq, 0.3)
                time.sleep(0.05)
            
            time.sleep(1.0)  # Longer pause between cycles
            
    except KeyboardInterrupt:
        print("\nStopping...")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        pipe_server.stop()
        print("Pipe server stopped.")
    
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: examples/CreatePipeExample.cpp
/*++

Copyright (c) Beehive Systems Pvt Ltd. All Rights Reserved

Module Name:

    CreatePipeExample.cpp

Abstract:

    Example demonstrating how to create a named pipe for the Virtual Audio Driver
    to connect to. This example shows the proper way to create and manage the
    named pipe that the kernel driver will open.

--*/

#include <windows.h>
#include <iostream>
#include <vector>
#include <cmath>

// Constants for audio generation
const int SAMPLE_RATE = 44100;
const int CHANNELS = 2;
const int BITS_PER_SAMPLE = 16;
const int BYTES_PER_SAMPLE = BITS_PER_SAMPLE / 8;
const int BUFFER_SIZE = SAMPLE_RATE * CHANNELS * BYTES_PER_SAMPLE / 10; // 100ms buffer

class VirtualMicPipeServer
{
private:
    HANDLE m_hPipe;
    std::wstring m_pipeName;
    bool m_isRunning;
    
public:
    VirtualMicPipeServer(int pinNumber = 0) : m_hPipe(INVALID_HANDLE_VALUE), m_isRunning(false)
    {
        // Create pipe name for the specified pin
        m_pipeName = L"\\\\.\\pipe\\VirtualMicInput_" + std::to_wstring(pinNumber);
    }
    
    ~VirtualMicPipeServer()
    {
        Stop();
    }
    
    bool Start()
    {
        // Create the named pipe
        m_hPipe = CreateNamedPipeW(
            m_pipeName.c_str(),
            PIPE_ACCESS_OUTBOUND,               // We write to the driver
            PIPE_TYPE_BYTE | PIPE_READMODE_BYTE | PIPE_WAIT,
            1,                                  // MaxInstances
            BUFFER_SIZE,                        // OutBufferSize  
            BUFFER_SIZE,                        // InBufferSize
            0,                                  // DefaultTimeOut
            NULL                                // Security attributes
        );
        
        if (m_hPipe == INVALID_HANDLE_VALUE)
        {
            std::wcerr << L"Failed to create named pipe: " << m_pipeName 
                      << L", Error: " << GetLastError() << std::endl;
            return false;
        }
        
        std::wcout << L"Named pipe created successfully: " << m_pipeName << std::endl;
        std::wcout << L"Virtual Audio Driver can now connect to this pipe." << std::endl;
        
        m_isRunning = true;
        return true;
    }
    
    void Stop()
    {
        m_isRunning = false;
        if (m_hPipe != INVALID_HANDLE_VALUE)
        {
            CloseHandle(m_hPipe);
            m_hPipe = INVALID_HANDLE_VALUE;
        }
    }
    
    bool SendSineWave(double frequency, double duration)
    {
        if (m_hPipe == INVALID_HANDLE_VALUE || !m_isRunning)
        {
            return false;
        }
        
        int totalSamples = (int)(SAMPLE_RATE * duration);
        std::vector<int16_t> buffer(totalSamples * CHANNELS);
        
        // Generate stereo sine wave
        for (int i = 0; i < totalSamples; i++)
        {
            double time = (double)i / SAMPLE_RATE;
            double sample = sin(2.0 * M_PI * frequency * time) * 16000; // Amplitude of 16000
            
            int16_t sampleValue = (int16_t)sample;
            buffer[i * 2] = sampleValue;     // Left channel
            buffer[i * 2 + 1] = sampleValue; // Right channel
        }
        
        // Send data to pipe
        DWORD bytesToWrite = totalSamples * CHANNELS * sizeof(int16_t);
        DWORD bytesWritten = 0;
        
        BOOL result = WriteFile(
            m_hPipe,
            buffer.data(),
            bytesToWrite,
            &bytesWritten,
            NULL
        );
        
        if (!result || bytesWritten != bytesToWrite)
        {
            std::wcerr << L"Failed to write audio data to pipe, Error: " 
                      << GetLastError() << std::endl;
            return false;
        }
        
        std::wcout << L"Sent " << bytesWritten << L" bytes (" << duration 
                  << L"s of " << frequency << L"Hz sine wave)" << std::endl;
        
        return true;
    }
    
    bool SendSilence(double duration)
    {
        if (m_hPipe == INVALID_HANDLE_VALUE || !m_isRunning)
        {
            return false;
        }
        
        int totalSamples = (int)(SAMPLE_RATE * duration);
        std::vector<int16_t> buffer(totalSamples * CHANNELS, 0); // All zeros for silence
        
        DWORD bytesToWrite = totalSamples * CHANNELS * sizeof(int16_t);
        DWORD bytesWritten = 0;
        
        BOOL result = WriteFile(
            m_hPipe,
            buffer.data(),
            bytesToWrite,
            &bytesWritten,
            NULL
        );
        
        if (!result || bytesWritten != bytesToWrite)
        {
            std::wcerr << L"Failed to write silence to pipe, Error: " 
                      << GetLastError() << std::endl;
            return false;
        }
        
        std::wcout << L"Sent " << duration << L"s of silence (" << bytesWritten 
                  << L" bytes)" << std::endl;
        
        return true;
    }
};

void PrintUsage()
{
    std::wcout << L"Usage: CreatePipeExample.exe [pin_number]" << std::endl;
    std::wcout << L"  pin_number: Virtual microphone pin number (default: 0)" << std::endl;
    std::wcout << L"" << std::endl;
    std::wcout << L"This program creates a named pipe and sends test audio data." << std::endl;
    std::wcout << L"Make sure the Virtual Audio Driver is installed and running." << std::endl;
}

int main(int argc, char* argv[])
{
    int pinNumber = 0;
    
    // Parse command line arguments
    if (argc > 1)
    {
        pinNumber = atoi(argv[1]);
        if (pinNumber < 0)
        {
            std::wcerr << L"Invalid pin number. Using default: 0" << std::endl;
            pinNumber = 0;
        }
    }
    
    if (argc > 1 && (strcmp(argv[1], "/?") == 0 || strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0))
    {
        PrintUsage();
        return 0;
    }
    
    std::wcout << L"Virtual Audio Driver - Named Pipe Example" << std::endl;
    std::wcout << L"=========================================" << std::endl;
    std::wcout << L"Using microphone pin: " << pinNumber << std::endl;
    std::wcout << L"" << std::endl;
    
    VirtualMicPipeServer pipeServer(pinNumber);
    
    // Start the pipe server
    if (!pipeServer.Start())
    {
        std::wcerr << L"Failed to start pipe server" << std::endl;
        return 1;
    }
    
    std::wcout << L"Pipe server started. Sending test audio..." << std::endl;
    std::wcout << L"Press Ctrl+C to stop." << std::endl;
    std::wcout << L"" << std::endl;
    
    try
    {
        // Send a series of test tones
        while (true)
        {
            // Send 1 second of 440Hz (A4)
            pipeServer.SendSineWave(440.0, 1.0);
            Sleep(100); // Small gap
            
            // Send 1 second of 880Hz (A5)  
            pipeServer.SendSineWave(880.0, 1.0);
            Sleep(100);
            
            // Send 1 second of silence
            pipeServer.SendSilence(1.0);
            Sleep(100);
            
            // Send 1 second of 1000Hz
            pipeServer.SendSineWave(1000.0, 1.0);
            Sleep(1000); // Longer pause between cycles
        }
    }
    catch (...)
    {
        std::wcout << L"Stopping..." << std::endl;
    }
    
    pipeServer.Stop();
    std::wcout << L"Pipe server stopped." << std::endl;
    
    return 0;
}